[chinpcbenjamin] iP

chinpcbenjamin opened the pull request:
# BenjaminBot

> "Your mind is for having ideas, not holding them." - David Allen

BenjaminBot frees your mind from having to remember your tasks. Here are some of its highlights:
- Text-based
- Easy to learn
- **SUPER FAST** to use ⚡


You can download BenjaminBot from [here](https://github.com/chinpcbenjamin/ip/releases/tag/A-Jar).

Features:
- [x] Managing Tasks
- [ ] Managing Deadlines
- [x] Managing Events

Why should **you** use it?
1. It's free!
2. It helps you manage your daily activities!
3. It is easy to learn!

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public static void main(String[] args) {
        BenjaminBot benjaminBot = new BenjaminBot(
                new Ui(),
                new TaskList(),
                new Storage("./data/benjamin.txt"));

        benjaminBot.ui.welcomeMessage();

        benjaminBot.storage.load(benjaminBot.taskArr);

        Parser parser = new Parser();
        Scanner sc = new Scanner(System.in);
        String s = sc.nextLine();
        benjaminBot.ui.printDivider();

        while (!s.isEmpty()) {
            parser.parse(
                    s,
                    benjaminBot.ui,
                    benjaminBot.taskArr,
                    benjaminBot);

            benjaminBot.ui.printDivider();
            s = sc.nextLine();
            benjaminBot.ui.printDivider();
        }
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

chinpcbenjamin committed:
Renamed Duke.java to BenjaminBot.java, and added Greet and Exit

chinpcbenjamin committed:
Implemented echo feature for BenjaminBot

chinpcbenjamin committed:
Added ability to store text enter by user and display back to user when requested

chinpcbenjamin committed:
Created Task class. Added ability to set Task as done and not done

chinpcbenjamin committed:
Added Todos, Events and Deadlines as subclass of Task

chinpcbenjamin committed:
Added error handling for some erroneous inputs

chinpcbenjamin committed:
Added ability to delete along with error handling. Added enum to handle different Add vs Delete actions.

chinpcbenjamin committed:
Add loading of tasks from .txt file

chinpcbenjamin committed:
Add saving of list of tasks to .txt file

chinpcbenjamin committed:
Merge branch 'branch-Level-7'

chinpcbenjamin committed:
Format any dates that users enter as a LocalDateTime instead of String

chinpcbenjamin committed:
Merge branch 'branch-Level-8'

chinpcbenjamin committed:
Create new Ui class to handle interactions with the user

chinpcbenjamin committed:
Create new TaskList class to keep track of the lists of tasks that is currently being used by BenjaminBot

chinpcbenjamin committed:
Create Storage class to handle reading and writing of saved list of tasks

chinpcbenjamin committed:
Created Parser class to handle the logic for decoding user commands

chinpcbenjamin committed:
Merge branch 'branch-A-MoreOOP'

chinpcbenjamin committed:
Put all classes into one 'BenjaminBot' Package

chinpcbenjamin committed:
Merge branch 'branch-A-Packages'

chinpcbenjamin committed:
Merge branch 'branch-A-Gradle'

chinpcbenjamin committed:
Add JUnit tests to test TaskList's addTask, removeTask and getTaskCount methods

chinpcbenjamin committed:
Merge branch 'branch-A-JUnit'

chinpcbenjamin committed:
Change build.gradle to create .jar file of the bot

chinpcbenjamin committed:
Merge branch 'branch-A-Jar'

chinpcbenjamin committed:
Add JavaDoc to BenjaminBot class

chinpcbenjamin committed:
Modify BenjaminBot class to be compliant with coding standard

chinpcbenjamin committed:
Add new condition in Parser class to handle 'find' commands

chinpcbenjamin committed:
Add JavaDoc comments for Task and Deadline classes

chinpcbenjamin committed:
Added JavaDoc for Event and Parser classes

chinpcbenjamin committed:
Add JavaDoc for Storage class

chinpcbenjamin committed:
Add JavaDoc for TaskList and Todo classes

chinpcbenjamin committed:
Added JavaDoc for Ui class

chinpcbenjamin committed:
Modify Deadline and Event classes to be compliant with coding standard.

chinpcbenjamin committed:
Modify Storage, Todo and Ui classes to be compliant with coding standard

chinpcbenjamin committed:
Add handling of find command to Ui, Task and TaskList classes

chinpcbenjamin committed:
Merge branch 'branch-A-JavaDoc'

chinpcbenjamin committed:
Merge branch 'branch-A-CodingStandard'

chinpcbenjamin committed:
Merge branch 'branch-Level-9'

chinpcbenjamin committed:
Add checkstyle

chinpcbenjamin committed:
Fix code to be in line with coding standard as tested by checkstyle

chinpcbenjamin committed:
Merge branch 'A-CheckStyle'

chinpcbenjamin committed:
Add application window to BenjaminBot

chinpcbenjamin committed:
Add dialog boxes and change code to display in gui

chinpcbenjamin committed:
Merge branch 'branch-Level-10'

advaypakhale commented on src/main/java/benjaminbot/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @Override
+    public String toString() {
+        return super.getIsDone()
+                ? "[D] [X] " + super.getTask() + " (by: "
+                    + this.endTime.format(DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm:ss")) + ")"
+
+                : "[D] [ ] " + super.getTask() + " (by: "
+                    + this.endTime.format(DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm:ss")) + ")";
+    }
```
maybe it is a better idea to avoid ternary statements here? it makes the code slightly harder to read, expanding this out into an if-else could make the code more readable

advaypakhale replied:
maybe you could expand this ternary statement out into an if-else to improve readability?

advaypakhale commented on src/main/java/benjaminbot/Parser.java:
```diff
@@ -0,0 +1,37 @@
+     * @param taskArr The TaskList instance that contains the current tasks of the BenjaminBot instance.
+     * @param benjaminBot The current instance of BenjaminBot that uses this parser.
+     */
+    public String parse(String s, Ui ui, TaskList taskArr, BenjaminBot benjaminBot) {
+        if (s.equals("list")) {
+            return ui.handleList(taskArr);
+        } else if (s.equals("bye")) {
+            return benjaminBot.exit();
+        } else if (s.startsWith("mark")) {
+            return ui.handleMark(s, taskArr);
+        } else if (s.startsWith("unmark")) {
+            return ui.handleUnmark(s, taskArr);
+        } else if (s.startsWith("todo")) {
+            return ui.handleTodo(s, taskArr);
+        } else if (s.startsWith("deadline")) {
+            return ui.handleDeadline(s, taskArr);
+        } else if (s.startsWith("event")) {
+            return ui.handleEvent(s, taskArr);
+        } else if (s.startsWith("delete")) {
+            return ui.handleDelete(s, taskArr);
+        } else if (s.startsWith("find")) {
+            return ui.handleFind(s, taskArr);
+        } else {
+            return ui.invalidCommandMessage();
+        }
+    }
```
might be cleaner if this were replaced with a switch case?

advaypakhale commented on src/main/java/benjaminbot/BenjaminBot.java:
```diff
@@ -0,0 +1,66 @@
+        return this.ui.welcomeMessage();
+    }
+
+    public String getResponse(String s) {
+        return parser.parse(
+                s,
+                this.ui,
+                this.taskArr,
+                this);
+    }
```
perhaps could use more descriptive parameter names?

advaypakhale commented on src/main/java/benjaminbot/TaskList.java:
```diff
@@ -0,0 +1,68 @@
+package benjaminbot;
```
could split into further packages, in order to increase modularity?

advaypakhale reviewed (commented):
Awesome stuff so far! Maybe you could consider splitting your code into more packages and making the code more readable in some places :)

kai824 commented on src/main/java/benjaminbot/BenjaminBot.java:
```diff
@@ -0,0 +1,66 @@
+    private Parser parser;
+
+    /**
+     * Constructs a BenjaminBot instance, given instances of Ui, TaskList and Storage.
+     * @param ui The user interface for the program to interact with the user.
+     * @param taskArr The list of tasks currently being stored.
+     * @param storage The storage object for loading and saving tasks.
```
Note the Javadoc formatting. This applies to your other Javadoc comments throughout
> Empty line between description and parameter section

kai824 commented on src/main/java/benjaminbot/BenjaminBot.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    public String getResponse(String s) {
+        return parser.parse(
+                s,
+                this.ui,
+                this.taskArr,
+                this);
```
Is there a particular reason for splitting them into multiple lines? It seems to easily fit in 80 characters if concatenated into 1 line

kai824 commented on src/main/java/benjaminbot/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public String toString() {
+        return super.getIsDone()
+                ? "[D] [X] " + super.getTask() + " (by: "
+                    + this.endTime.format(DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm:ss")) + ")"
+
+                : "[D] [ ] " + super.getTask() + " (by: "
+                    + this.endTime.format(DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm:ss")) + ")";
```
Considering that both cases are almost equal other than the first string, consider using the tertiary statement to handle only the first part (`[D] [X]`). Something like:
```java
return (super.getIsdone() ? "[D] [X] " : "[D] [ ]") + super.getTask() + ...
```

Moreover, given that other child classes of Task also have this `[X]` and `[ ]` checkbox, consider creating a new method in Task that returns the checkbox together with the Task string. This way, you can avoid having this check in all child classes of Task.

kai824 commented on src/main/java/benjaminbot/Storage.java:
```diff
@@ -0,0 +1,97 @@
+     */
+    private void loadSavedTasks(String s, TaskList arr) {
+        String[] stringArray = s.split(",");
+        switch (stringArray[0]) {
+        case "T":
+            arr.addTask(new Todo(
+                    stringArray[2],
+                    stringArray[1].equals("1")));
+            break;
+        case "D":
+            arr.addTask(new Deadline(
+                    stringArray[2],
+                    stringArray[1].equals("1"),
+                    LocalDateTime.parse(stringArray[3])));
+            break;
+        case "E":
+            arr.addTask(new Event(stringArray[2],
+                    stringArray[1].equals("1"),
+                    LocalDateTime.parse(stringArray[3]),
+                    LocalDateTime.parse(stringArray[4])));
+            break;
+        default:
+            break;
+        }
```
Good job in setting up your IDE to follow the switch-case indentation of this course!

kai824 commented on src/main/java/benjaminbot/Task.java:
```diff
@@ -0,0 +1,66 @@
+     * @param s The description of the task.
+     * @param b A boolean describing whether the task has been completed or not.
+     */
+    public Task(String s, boolean b) {
+        this.task = s;
+        this.isDone = b;
+    }
```
Consider using more descriptive parameter names

kai824 commented on src/main/java/benjaminbot/Todo.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    /**
+     *  Constructs a todo instance specified by the string description. This task is not completed yet.
```
Just a minor issue, but note the leading space before `Constructs`

kai824 reviewed (commented):
Overall, the code is well-written and well-documented!

However, do take note of the javadoc formatting. Remember to leave a blank line for all javadocs before the first `@param` argument

chinpcbenjamin committed:
Update gradle configuration file

The previous build.gradle was misconfigured, and thus did not allow for
building of a .jar file.

To ensure that a .jar file of the application can be built in the
future, let's fix the configuration file.

chinpcbenjamin committed:
Add assert check to TaskList addTask method

The TaskList should not have tasks that are null.

Let's add an assertion in the addTask method to check that any task we
attempt to add to a TaskList is indeed not null.

chinpcbenjamin committed:
Add assertions to various Ui commands

There are several different methods in the Ui class that are called by
the Parser. There is a chance that the wrong method is accidentally
called.

To detect mistakes in the event that the parse command is updated, let
us make some assertions about what the string command was to ensure
that the right method is called by the parser.

chinpcbenjamin committed:
Remove dead code

There were several lines of unused code that remained. These were
originally used, but when the code was modified such that it became
unused, it was not deleted.

In order to maintain good code quality, dead code was removed from the
BenjaminBot and MainWindow classes.

chinpcbenjamin committed:
Add lines between Javadoc description and
parameter section to ensure correct Javadoc format

chinpcbenjamin committed:
Refactor toString() of Task and its subclass

To reduce arrowheads in the code and to take advantage of similar lines
of code in the various toString() methods of the Task class and its
subclasses, the code was refactored. This improves the code quality by
making it easier to read.

chinpcbenjamin committed:
Merge pull request #1 from chinpcbenjamin/branch-A-Assertions

Add assertions to code

chinpcbenjamin committed:
Merge branch 'master' into branch-A-CodeQuality

chinpcbenjamin committed:
Merge pull request #2 from chinpcbenjamin/branch-A-CodeQuality

Improve Code Quality

chinpcbenjamin committed:
Add functions to Parser, Task, TaskList and Ui to
support adding view schedule functionality

chinpcbenjamin committed:
Implement isSameDate in subclasses of Task class

chinpcbenjamin committed:
Add complete view schedule functionality

Change methods in Ui, TaskList, and Task and its subclasses to make use
of LocalDate instead of LocalDateTime, since the schedule shows all the
tasks on a specific day and thus LocalDateTime is not necessary.

chinpcbenjamin committed:
Merge branch 'branch-B-ViewSchedules'

chinpcbenjamin committed:
Update README.md

chinpcbenjamin committed:
Merge branch 'branch-A-UserGuide'

chinpcbenjamin committed:
Add images representing bot and user

chinpcbenjamin committed:
Add icon and title to stage

chinpcbenjamin committed:
Add icon and title to stage

chinpcbenjamin committed:
Merge branch 'branch-images'

chinpcbenjamin committed:
Update Javadoc

chinpcbenjamin committed:
Merge branch 'branch-more-javadoc'

chinpcbenjamin committed:
Fix view and list commands

Updated the view and list commands to feedback to the user if
there are no tasks that available to return.

chinpcbenjamin committed:
Merge branch 'branch-list-and-view-fixes'

chinpcbenjamin committed:
Update README.md
